internal/queue/queue.go

```diff
@@ -198,15 +231,16 @@ func (q *Queue) insert(ctx context.Context, key string, ratelimit bool, delay ti
 	}
 
 	if ratelimit {
-		if delay > 0 {
-			panic("Non-zero delay with rate limiting not supported")
+		actualDelay := q.ratelimiter.When(key)
+		// Check if delay is overridden
+		if delay != nil {
+			actualDelay = *delay
 		}
-		ratelimitDelay := q.ratelimiter.When(key)
-		span.WithField(ctx, "delay", ratelimitDelay.String())
-		val.plannedToStartWorkAt = val.plannedToStartWorkAt.Add(ratelimitDelay)
-		val.delayedViaRateLimit = &ratelimitDelay
+		span.WithField(ctx, "delay", actualDelay.String())
```

Comment:
We should special case stringers in the tracing so it only does the conversion when we are actually collecting the trace.

Reply:
How do you think we should special case nil duration? or should nil always just be nil

Reply:
We could say use `-1` (or < 0) instead of `nil` to avoid all these pointer conversions.

Reply:
There's already code (And tests for it), that allow you to pass a negative delay from the rate limiter back and result in it being processed immediately. That would regress / change this behaviour.

Reply:
I don't understand why we did that since 0 should mean no delay. But it is what it is.

Reply:
It's the way the kubernetes ratelimiters work, and we kept the contract the same.